File: p3/main.java
package p3;

import java.util.function.Consumer;
import java.util.function.IntConsumer;

enum Command {
    FIZZ, BUZZ, FIZZBUZZ, NUMBER
}

class FizzBuzz {
    private int n, i;

    public FizzBuzz(int n) {
        this.n = n;
        this.i = 1;
    }

    private void oneStep(Consumer<Object> action, Command command) throws InterruptedException {
        while(true) {
            synchronized(this){
                if(i > n) {
                    notify();
                    return;
                }
                while((command == Command.FIZZBUZZ && i % 15 != 0) ||
                        (command == Command.FIZZ && (i % 3 != 0 || i % 5 == 0)) ||
                        (command == Command.BUZZ && (i % 5 != 0 || i % 3 == 0)) ||
                        (command == Command.NUMBER && (i % 3 == 0 || i % 5 == 0))){
                    wait();
                    if(i > n) {
                        notify();
                        return;
                    }
                }
            } 
            action.accept(i);
            synchronized(this){
                i++;
                notifyAll();
            }
        }
    }

    public void fizz(Runnable printFizz) throws InterruptedException {
            oneStep(obj -> printFizz.run(), Command.FIZZ);
    }

    public void buzz(Runnable printBuzz) throws InterruptedException {
        oneStep(obj -> printBuzz.run(), Command.BUZZ);
    }

    public void fizzbuzz(Runnable printFizzBuzz) throws InterruptedException {
        oneStep(obj -> printFizzBuzz.run(), Command.FIZZBUZZ);
    }

    public void number(IntConsumer printNumber) throws InterruptedException {
        oneStep(obj -> printNumber.accept((Integer)obj), Command.NUMBER);
    }
}

File: p3/main.cpp
#include <condition_variable>
class FizzBuzz {
private:
    int n, i;
    std::mutex m;
    std::condition_variable cv;

    void oneStep(function<void(int)> printFunc, function<bool(int)> condition){
        while(true){
            {
                std::unique_lock lk(m);
                if(i > n){
                    cv.notify_one();
                    return;
                }
                if(!condition(i)){
                    cv.wait(lk);
                    continue; // this is probably less performant than adding a second i > n check here
                }
            }
            printFunc(i);
            {
                std::unique_lock lk(m);
                i++;
            }
            cv.notify_all();
        }
    }

public:
    FizzBuzz(int n) {
        this->n = n;
        this->i = 1;
    }

    void fizz(function<void()> printFizz) {
        oneStep([printFizz](int x){return printFizz();}, [](int x){return x % 3 == 0 && x % 5 != 0;});
    }

    void buzz(function<void()> printBuzz) {
        oneStep([printBuzz](int x){return printBuzz();}, [](int x){return x % 5 == 0 && x % 3 != 0;});
    }

	void fizzbuzz(function<void()> printFizzBuzz) {
        oneStep([printFizzBuzz](int x){return printFizzBuzz();}, [](int x){return x % 15 == 0;});
    }

    void number(function<void(int)> printNumber) {
        oneStep(printNumber, [](int x){return x % 3 != 0 && x % 5 != 0;});
    }
};